source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,209 @@
+  return level > 0 ? level : DEFAULT_MEMORY_LEVEL;
+}
+
+uint64_t GzipFilterConfig::windowBitsUint(Protobuf::uint32 window_bits) {
+  return (window_bits > 0 ? window_bits : DEFAULT_WINDOW_BITS) | GZIP_HEADER_VALUE;
+}
+
+GzipFilter::GzipFilter(const GzipFilterConfigSharedPtr& config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available content-encoding in Envoy at the
+  // moment, order/priority of preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ =
+      !(headers.AcceptEncoding() &&
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    // TODO(gsagula): tests needed.
+    config_->disableVary()
+        ? headers.removeVary()
+        : headers.insertVary().value(Http::Headers::get().VaryValues.AcceptEncoding);
+
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length{};
+    StringUtil::atoul(content_length->value().c_str(), length);
+    return length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding && transfer_encoding->value().find(
+                                   Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
```

Comment:
again I'd probably tokenize, trim, and do an explicit lookup for "chunked".  This would match "chunkedfoo" which probably never happens, but still :)

Reply:
Sorry, I and my "find" methods... I will review all the other ones as well!